app/models/saved_page.rb

```diff
@@ -0,0 +1,32 @@
+class SavedPage < ApplicationRecord
+  belongs_to :oidc_user
+
+  validates :page_path, uniqueness: { scope: :oidc_user_id }, presence: true
+
+  validate :page_path_is_valid_path
+  validate :page_path_is_url_safe_string
+
+private
+
+  def page_path_is_url_safe_string
+    errors.add(:page_path, :url_unsafe, message: "must be a URL safe string") if is_url_unsafe_string?
+  end
+
+  def page_path_is_valid_path
+    errors.add(:page_path, :invalid_path, message: "must only include URL path and fragment") if is_invalid_path?
+  end
+
+  def is_invalid_path?
+    return true if page_path.blank?
+
+    uri = URI(CGI.unescape(page_path))
+    fragment = "##{uri.fragment}" if uri.fragment
```

Comment:
I don't think we should have fragments, it should only be a path.

With fragments, we're implementing a "save this section of this page" feature and not a "save this page" feature.

Reply:
Humm yeah I see what you mean, though we will control which link we hand to the button to persist right?

Mostly what I was thinking about with this was the `/bank-holiday#scotland` URL, where a user may want to save that specific page, but to lose the framgent gets them to different content?

Reply:
I think it's a lesser can of worms to handle the bank holidays page with the user's saved location, and to keep this feature for saving pages (and not parts of pages)

Reply:
Ok cool I'll scrub that bit